src/integrations/devicecontext.ts

```diff
@@ -19,29 +20,127 @@ export class DeviceContext implements Integration {
+        );
+      }
+
+      if (!event.level && nativeContexts.level) {
+        event.level = severityLevelFromString(nativeContexts.level);
+      }
+
+      const nativeEnvironment = nativeContexts.environment;
+      if (!event.environment && nativeEnvironment) {
+        event.environment = nativeEnvironment;
+      }
+
+      const nativeBreadcrumbs = Array.isArray(nativeContexts.breadcrumbs)
+      ? nativeContexts.breadcrumbs.map(breadcrumbFromObject)
+      : undefined;
+      if (nativeBreadcrumbs) {
+        if (event.breadcrumbs && event.breadcrumbs.length !== nativeBreadcrumbs.length) {
+          const maxBreadcrumbs = hub.getClient()?.getOptions().maxBreadcrumbs ?? 100; // Default is 100.
+          event.breadcrumbs = this._mergeUniqueBreadcrumbs(event.breadcrumbs, nativeBreadcrumbs, maxBreadcrumbs);
+        }
+        else {
+          event.breadcrumbs = nativeBreadcrumbs;
+        }
+      }
 
       return event;
     });
   }
+
+  /**
+ * Merges two groups of ordered breadcrumbs and removes any duplication that may
+ * happen between them.
+ * @param jsList The first group of breadcrumbs from the JavaScript layer.
+ * @param nativeList The second group of breadcrumbs from the native layer.
+ * @returns An array of unique breadcrumbs merged from both lists.
+   */
+  private _mergeUniqueBreadcrumbs(jsList: Breadcrumb[], nativeList: Breadcrumb[], maxBreadcrumbs: number): Breadcrumb[] {
+    // Ensure both lists are ordered by timestamp.
+    const orderedNativeList = [...nativeList].sort((a, b) => (a.timestamp ?? 0) - (b.timestamp ?? 0));
+    const orderedJsList = [...jsList].sort((a, b) => (a.timestamp ?? 0) - (b.timestamp ?? 0));
+
+    const combinedList: Breadcrumb[] = [];
+    let jsIndex = 0;
+    let natIndex = 0;
+
+    while (jsIndex < orderedJsList.length && natIndex < orderedNativeList.length && combinedList.length < maxBreadcrumbs)
+    {
+      const jsBreadcrumb = orderedJsList[jsIndex];
+      const natBreadcrumb = orderedNativeList[natIndex];
+
+      if (jsBreadcrumb.timestamp === natBreadcrumb.timestamp &&
+        jsBreadcrumb.message === natBreadcrumb.message) {
+        combinedList.push(jsBreadcrumb);
+        jsIndex++;
+        natIndex++;
+      }
+      else if (jsBreadcrumb.timestamp && natBreadcrumb.timestamp &&
+        jsBreadcrumb.timestamp < natBreadcrumb.timestamp)
+      {
+        combinedList.push(jsBreadcrumb);
+        jsIndex++;
+      }
+      else {
+        combinedList.push(natBreadcrumb);
+        natIndex++;
+      }
+    }
+
+    // Add remaining breadcrumbs from the JavaScript and Native list if space allows.
+    while (jsIndex < orderedJsList.length  && combinedList.length < maxBreadcrumbs) {
+      combinedList.push(orderedJsList[jsIndex++]);
+    }
+
+    while (natIndex < orderedNativeList.length && combinedList.length < maxBreadcrumbs) {
+      combinedList.push(orderedNativeList[natIndex++]);
+    }
+    return combinedList;
+  }
```

Comment:
The function works, but I think we should consider marking the JS crumbs during the creation on native layers, as then we can also filter then on the native layers and only return the native crumbs.

The native solution would be faster and required less memory.

We could do `JSBreadcrumb extends Breadcrumb` and then filter `breadcrumb instanceof JSBreadcrumb`.

Reply:
yeah, let's please not use `unknown` - it exists to work-around JVM's static type system exclusively. A new field is the cleanest way imo, but we could also open the `Breadcrumb` class for extension, should be not a problem

Reply:
I also like the idea of breadcrumb.origin, it could also help in the future to know if a breadcrumb was originated from the JavaScript layer or native

Reply:
> I also like the idea of breadcrumb.origin, it could also help in the future to know if a breadcrumb was originated from the JavaScript layer or native

Hey @lucas-zimerman 👋 
I've [drafted an iOS PR](https://github.com/getsentry/sentry-cocoa/pull/4358/files) and was wondering if you have any early feedback on it. Does the approach chosen unblock this PR?

Reply:
It surelly does thank you!